client/txn.go

```diff
@@ -300,18 +301,31 @@ func (txn *Txn) CommitInBatch(b *Batch) error {
 
 // Commit sends an EndTransactionRequest with Commit=true.
 func (txn *Txn) Commit() error {
-	if txn.Proto.Writing {
-		return txn.sendEndTxnCall(true /* commit */)
+	err := txn.sendEndTxnCall(true /* commit */)
+	if err == nil {
+		return nil
 	}
-	return nil
+	if !txn.execing {
+		// If this transaction is not part of the auto-retry loop we need to
+		// cleanup the transaction on retryable commit failures.
+		if _, ok := err.(proto.TransactionRestartError); ok {
+			// Always allow the transaction abort request to be sent, even if we have
+			// already sent/queued an EndTransaction request.
+			txn.haveEndTxn = false
```

Comment:
Since SQL transactions do not go through the `exec()` I believe it's not possible for `haveEndTxn` to be set at all.